### ISSUE TITLE
Index signature not checked when assigning value in constructor

### ISSUE THREAD

**Issue by dbartholomae** [microsoft/TypeScript#53950]
# Bug Report

### 🔎 Search Terms

index signature constructor

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about constructors

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgJLIN4Chm+TAewIC5Mc8KBtAawgE9SBnMKUAcwF1SQBXAWwBG0ANzlcAXyySsCADZxGjZAGFkwPgAdZEPhHBL02CoRKZJFBARDMoPBGAJRkACgCUZChTAALYIwB0JsgAvJjIAnBQpABE0cjmeJKSQA)

### 💻 Code

```ts
interface I {
    foo: {
        [key: string]: number;
    }
}

class C implements I {
    foo: {}
    constructor () {
        // This should be an error
        this.foo = { bar: "" }
    }
}
```

### 🙁 Actual behavior

No TypeScript error

### 🙂 Expected behavior

I would have expected to get a type error because `this.foo.bar` was assigned a string, but the interface I that is implemented by C allows only `number`.

**Comment by dbartholomae** [comment]
Thanks! I was also pointed to this part of the documentation which explains it:
https://www.typescriptlang.org/docs/handbook/2/classes.html#cautions

**Comment by MartinJohns** [comment]
See also: #32082
